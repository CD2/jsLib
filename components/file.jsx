import React from "react"
import PropTypes from "prop-types"
import { buildUrl } from "lib/utils/api_http"
import { styled } from "lib/utils/theme"
import decorate from "lib/utils/decorate"
import Button from "./button"
import FaIcon from "./fa_icon"

export class File extends React.Component {
  static propTypes = {
    children: PropTypes.any,
    className: PropTypes.string,
    defaultSrc: PropTypes.string,
    linkOnly: PropTypes.bool,
    linkTarget: PropTypes.string,
    onClick: PropTypes.func,
    uid: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
    wide: PropTypes.bool,
  }

  static defaultProps = {
    alt: ``,
    background: false,
    crop: false,
    wide: false,
  }

  get url() {
    const { uid } = this.props
    const params = { uid }
    return buildUrl([`/file`], params)
  }

  render() {
    const { defaultSrc, uid, children, linkOnly, linkTarget } = this.props

    let url = this.url
    if (!uid) {
      url = defaultSrc
    }
    if (linkOnly) {
      return (
        <a href={url} target={linkTarget ? linkTarget : `_blank`}>
          {children}
        </a>
      )
    }
    return (
      <Button
        to={linkTarget ? linkTarget : url}
        className={`file ${this.props.className}`}
        target="_blank"
        buttonStyle="download common"
        wide={this.props.wide}
        onClick={this.props.onClick}
      >
        <FaIcon icon="download" size={1.3} />
        {children}
      </Button>
    )
  }
}
export default decorate(
  styled`
    color: white;
    font-weight: 600;
    white-space: nowrap;
    i {
      margin-right: 6px;
      color: white;
    }
  `,
  File,
)
