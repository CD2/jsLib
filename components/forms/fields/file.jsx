
import React from 'react'
import PropTypes from 'prop-types'
import { observable, action } from 'mobx'
import { observer } from 'mobx-react'
import FileDownload from "lib/components/file"

@observer
export class FileField extends React.Component {

  static propTypes = {
    accepts: PropTypes.array,
    file_name: PropTypes.string,
    multiple: PropTypes.bool,
    name: PropTypes.string,
    onChange: PropTypes.func,
    onFocus: PropTypes.func,
    onRawChange: PropTypes.func,
    onlySpreadsheets: PropTypes.bool,
    uid: PropTypes.string,
    value: PropTypes.object,
  }

  static defaultProps = {
    accepts: [],
    onlyImages: false,
    onlySpreadsheets: false,
  }


  componentWillReceiveProps(props) {
    if (props.value === null && this.props.value) this.fileInput.value = ``
    if (!props.value) return
    if (props.value instanceof File) {
      let reader = new FileReader()
      reader.onload = action((e) => this.preview_src = e.target.result)
      reader.readAsDataURL(props.value)
    } else {
      this.preview_src = props.value
    }
  }

  @observable preview_src = ``

  handleChange = (e) => {
    const { onRawChange, onChange } = this.props
    if (onRawChange) onRawChange(e)
    if (onChange) {
      const file = e.target.files[0] || {}

      onChange({
        name: e.target.name,
        value: file,
        filename: file.name,
        size: file.size,
        type: file.type,
      })
    }
  }

  render() {
    const { name, multiple, onFocus, onlySpreadsheets } = this.props
    let acceptedTypes = null

    if (onlySpreadsheets) {
      acceptedTypes = `.xls,.xlw,.xlt,.xml,.xlsx,.xlsm,.xltx,.xltm,.xlsb`
    }
    return (
      <div>
        <input
          ref={element => this.fileInput = element}
          type="file"
          name={name}
          multiple={multiple}
          accept={acceptedTypes || this.props.accepts.join()}
          onChange={this.handleChange}
          onFocus={onFocus}
        />
        {this.props.uid ? <FileDownload uid={this.props.uid}>
          {this.props.file_name ? this.props.file_name : `Download`}
        </FileDownload> : null}
      </div>
    )
  }

}
export default FileField
