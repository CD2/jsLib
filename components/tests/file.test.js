/* eslint-disable react/jsx-filename-extension */
import React from "react"
import { File } from "../file"
import { shallow } from "enzyme"
import toJson from "enzyme-to-json"

describe(`<File />`, () => {
  const view = shallow(<File />)

  it(`snapshot`, () => expect(toJson(view)).toMatchSnapshot())
})
