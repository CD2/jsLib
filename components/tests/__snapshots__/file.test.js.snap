// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<File /> snapshot 1`] = `
<withTheme(StyledComponent)
  buttonStyle="download common"
  className="file undefined"
  target="_blank"
>
  <withTheme(StyledComponent)
    icon="download"
    size={1.3}
  />
</withTheme(StyledComponent)>
`;
